go/cli/mcap/cmd/recover.go

```diff
@@ -0,0 +1,326 @@
+package cmd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/foxglove/mcap/go/cli/mcap/utils"
+	"github.com/foxglove/mcap/go/mcap"
+	"github.com/spf13/cobra"
+)
+
+type recoverOptions struct {
+	decodeChunk bool
+	chunkSize   int64
+	compression mcap.CompressionFormat
+}
+
+func recoverRun(
+	r io.Reader,
+	w io.Writer,
+	ops *recoverOptions,
+) error {
+	decodeChunk := ops.decodeChunk
+	mcapWriter, err := mcap.NewWriter(w, &mcap.WriterOptions{
+		Chunked:     true,
+		ChunkSize:   ops.chunkSize,
+		Compression: ops.compression,
+	})
+	if err != nil {
+		return err
+	}
+
+	info := &mcap.Info{
+		Statistics: &mcap.Statistics{
+			ChannelMessageCounts: make(map[uint16]uint64),
+		},
+		Channels: make(map[uint16]*mcap.Channel),
+		Schemas:  make(map[uint16]*mcap.Schema),
+	}
+
+	defer func() {
```

Comment:
i don't think this makes sense to do in a defer - you only really want to do this in the success case when you hit DataEnd.